zilencer/views.py

```diff
@@ -158,33 +162,52 @@ def remote_server_notify_push(request: HttpRequest, entity: Union[UserProfile, R
 @zulip_login_required
 def initial_upgrade(request: HttpRequest) -> HttpResponse:
     user = request.user
+    error_message = ""
+
     if Customer.objects.filter(realm=user.realm).exists():
         return HttpResponseRedirect(reverse('zilencer.views.billing_home'))
 
     if request.method == 'POST':
-        stripe_customer = do_create_customer_with_payment_source(user, request.POST['stripeToken'])
-        # TODO: the current way this is done is subject to tampering by the user.
-        seat_count = int(request.POST['seat_count'])
-        if seat_count < 1:
-            raise AssertionError('seat_count is less than 1')
-        do_subscribe_customer_to_plan(
-            stripe_customer=stripe_customer,
-            stripe_plan_id=Plan.objects.get(nickname=request.POST['plan']).stripe_plan_id,
-            seat_count=seat_count,
-            # TODO: billing address details are passed to us in the request;
-            # use that to calculate taxes.
-            tax_percent=0)
-        # TODO: check for errors and raise/send to frontend
-        return HttpResponseRedirect(reverse('zilencer.views.billing_home'))
+        signed_seat_count = request.POST['signed_seat_count']
+        salt = request.POST['salt']
+
+        try:
+            seat_count = int(unsign_string(signed_seat_count, salt))
+            if seat_count < 1:
+                error_message = "Seat count is less than one."
+        except signing.BadSignature:
+            billing_logger.warning("Tampered seat count during realm upgrade: %s(%s)",
+                                   user.email, user.realm.string_id)
+            error_message = "Something went wrong. Please contact support@zulipchat.com"
+
+        if not error_message:
```

Comment:
hm, is there any way to write this so we just raise an error (e.g. JsonableError) rather than doing a `if not error` pattern? What do we do for other forms in the codebase?

Reply:
We can do a JsonableError if we create a separate view only for the POST request which returns back a JSON sucess response. The redirection to the billing page would be done by the JavaScript. There wont be any changes to GET request.